src/storage/mod.rs

```diff
@@ -9,16 +10,52 @@ use crate::{db::Pool, Config, Metrics};
 use chrono::{DateTime, Utc};
 use failure::{err_msg, Error};
 use path_slash::PathExt;
+use serde::{Deserialize, Serialize};
 use std::{
     collections::{HashMap, HashSet},
     ffi::OsStr,
     fmt, fs,
+    io::{self, Write},
+    ops::RangeInclusive,
     path::{Path, PathBuf},
     sync::Arc,
 };
 
 const MAX_CONCURRENT_UPLOADS: usize = 1000;
 
+#[derive(Clone, Debug, PartialEq, Serialize, Deserialize)]
+pub struct FileRange {
+    inner: RangeInclusive<u64>,
+    len: u64,
+}
+
+impl FileRange {
+    pub fn new(start: u64, end: u64) -> Self {
+        Self {
+            inner: start..=end,
+            len: end - start + 1,
+        }
+    }
+    pub fn start(&self) -> &u64 {
+        self.inner.start()
+    }
+    pub fn end(&self) -> &u64 {
+        self.inner.end()
+    }
+    pub fn len(&self) -> &u64 {
+        &self.len
+    }
```

Comment:
nit: since this is always based on `inner`, can you calculate it lazily instead? Then it can't get out of sync.
```suggestion
    pub fn len(&self) -> &u64 {
        self.inner.end - self.inner.start + 1
    }
```
That said, this type is a bit overkill - I think just using `end - start + 1` at the call site would be clear enough.

Reply:
@jyn514 the UB with the previous range-type (when iterating it) made me wrap it, but I can see that this is probably overkill :)